From: Liam Howlett <liam.howlett@oracle.com>
Subject: mm/mmap: fix do_brk_munmap() when munmapping multiple mappings
Date: Thu, 19 May 2022 15:05:22 +0000

If the newbrk is less than the last brk vma start, then munmap entire
VMA(s).

Link: https://lkml.kernel.org/r/20220519150509.1290067-1-Liam.Howlett@oracle.com
Signed-off-by: Liam R. Howlett <Liam.Howlett@oracle.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/mmap.c |    3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

--- a/mm/mmap.c~mm-mmap-change-do_brk_munmap-to-use-do_mas_align_munmap-fix
+++ a/mm/mmap.c
@@ -3076,8 +3076,7 @@ static int do_brk_munmap(struct ma_state
 
 	arch_unmap(mm, newbrk, oldbrk);
 
-	if (likely((vma->vm_end < oldbrk) ||
-		   ((vma->vm_start == newbrk) && (vma->vm_end == oldbrk)))) {
+	if (likely((vma->vm_end < oldbrk) || (vma->vm_start >= newbrk))) {
 		/* remove entire mapping(s) */
 		ret = do_mas_align_munmap(mas, vma, mm, newbrk, oldbrk, uf,
 					  true);
_
